pkg/controller/miganalytic/miganalytics_controller.go

```diff
@@ -0,0 +1,582 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package miganalytic
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/konveyor/controller/pkg/logging"
+	"github.com/konveyor/mig-controller/pkg/apis/migration/v1alpha1"
+	"github.com/konveyor/mig-controller/pkg/compat"
+	"github.com/konveyor/mig-controller/pkg/gvk"
+	"github.com/openshift/api/image/docker10"
+	"github.com/openshift/library-go/pkg/image/reference"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/dynamic"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	migapi "github.com/konveyor/mig-controller/pkg/apis/migration/v1alpha1"
+	imagev1 "github.com/openshift/api/image/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	k8sclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	MiBSuffix = "Mi"
+	MiB       = 1048576
+)
+
+var log = logging.WithName("analytics")
+
+// Add creates a new MigAnalytic Controller and adds it to the Manager with default RBAC. The Manager will set fields on the Controller
+// and Start it when the Manager is Started.
+func Add(mgr manager.Manager) error {
+	return add(mgr, newReconciler(mgr))
+}
+
+// newReconciler returns a new reconcile.Reconciler
+func newReconciler(mgr manager.Manager) reconcile.Reconciler {
+	return &ReconcileMigAnalytic{Client: mgr.GetClient(), scheme: mgr.GetScheme()}
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New("miganalytic-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		log.Trace(err)
+		return err
+	}
+
+	// Watch for changes to MigAnalytic
+	err = c.Watch(
+		&source.Kind{Type: &migapi.MigAnalytic{}},
+		&handler.EnqueueRequestForObject{},
+		&AnalyticPredicate{})
+	if err != nil {
+		log.Trace(err)
+		return err
+	}
+
+	return nil
+}
+
+var _ reconcile.Reconciler = &ReconcileMigAnalytic{}
+
+// ReconcileMigAnalytic reconciles a MigAnalytic object
+type ReconcileMigAnalytic struct {
+	client.Client
+	scheme *runtime.Scheme
+}
+
+func (r *ReconcileMigAnalytic) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	var err error
+	log.Reset()
+
+	// Fetch the MigAnalytic instance
+	analytic := &migapi.MigAnalytic{
+		Spec: migapi.MigAnalyticSpec{
+			MaxListImages: 25,
```

Comment:
I would recommend:


1. moving "25" into a named const so we don't have magic numbers, and so that we can change this in one place if changes are needed in future
2. not defaulting in this way, let's keep consistent across our `Get` operations and just initialize an empty object before running `Get`

I feel that this is an unusual way to apply a default. I would recommend just grabbing the const value you create from wherever you need it later on instead  of baking it into the object as a default.

Reply:
Agree with the constant. I missed that while cleaning up. It's also not the '50' I put in the docs. If you enable listImages and there isn't a default value set like this the default will be 0. And how do you differentiate a user set value of 0 and a default of 0. Reading about how to set non-zero defaults in Go this is the pattern that was described.

Reply:
Moved to a constant. Still want to set a non-zero default. This is the only way I know how unless someone has a better idea.

Reply:
This isn't working right if the value is absent from the CR completely. Removed it.